Source/WebGPU/WebGPU/CommandEncoder.mm

```diff
@@ -152,26 +158,28 @@ static MTLStoreAction storeAction(WGPUStoreOp storeOp)
             attachment.clearColor.b,
             attachment.clearColor.a);
 
+        addTextureStorageIfNeeded(fromAPI(attachment.view), attachment.resolveTarget ? WGPUStoreOp_Discard : attachment.storeOp, attachment.loadOp);
```

Comment:
We have to be really careful here. It's possible to submit commands in a different order than they were recorded. This code executes on the recording timeline, not the submission timeline. Naively, the logic about "X happens before Y" needs to be done on the submission timeline, because that's the order the GPU sees them.

I think the current patch's implementation of `addTextureStorageIfNeeded()` coincidentally solves this: Imagine there's a command buffer that is recorded using the memoryless texture, but the texture gets backing store allocated (and maybe even used!) in the window of time between our command buffer getting created and submitted. If our command buffer neither reads nor writes to the texture (which is the current logic of `addTextureStorageIfNeeded()`), then it's sort of a benign unobservable error that we recorded the command buffer referencing the wrong Metal texture.

On the other hand, the optimization only works if content _neither reads nor writes_ to the texture, which seems like a pattern I wouldn't expect to be common; the application could just ... not bind the render target at all, and have the same behavior.

If we tried to make the optimization more broadly applicable, and modified `addTextureStorageIfNeeded()` to only consider the store op, but not the load op, then that's incorrect because of the reordering problem described above.

Reply:
Exactly. The optimization is primarily for MSAA where you need the samples but can resolve the 4x (for example) samples -> 1x prior to committing any stores. Also with MSAA, depth sample count must match color sample count, so it is possible to eliminate the 4x 32bpp depth buffer memory too.

A non-MSAA optimization is for single pass rendering: you can eliminate the storage of the depth buffer due to how TBDR work. A depth attachment is still needed however but it seems like it's mostly for validation.